### PR TITLE
Updated user account objects to use NSSecureCoding

### PR DESCRIPTION
**Pull request by MikeNachbaur**
This hardens these objects to ensure the values returned are indeed the types we expect them to be.

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m [hunk 1]
```diff
@@ -117,22 +122,24 @@ - (void)encodeWithCoder:(NSCoder*)encoder {
...
+        _accessScopes     = [decoder decodeObjectOfClass:[NSSet class] forKey:kUser_ACCESS_SCOPES];
+        _email            = [decoder decodeObjectOfClass:[NSString class] forKey:kUser_EMAIL];
+        _fullName         = [decoder decodeObjectOfClass:[NSString class] forKey:kUser_FULL_NAME];
+        _credentials      = [decoder decodeObjectOfClass:[SFOAuthCredentials class] forKey:kUser_CREDENTIALS];
+        _idData           = [decoder decodeObjectOfClass:[SFIdentityData class] forKey:kUser_ID_DATA];
+        _organizationName = [decoder decodeObjectOfClass:[NSString class] forKey:kUser_ORGANIZATION_NAME];
+        _userName         = [decoder decodeObjectOfClass:[NSString class] forKey:kUser_USER_NAME];
+        _communityId      = [decoder decodeObjectOfClass:[NSString class] forKey:kUser_COMMUNITY_ID];
+        _communities      = [decoder decodeObjectOfClass:[NSArray class] forKey:kUser_COMMUNITIES];
+        _customData       = [[decoder decodeObjectOfClass:[NSDictionary class] forKey:kUser_CUSTOM_DATA] mutableCopy];
```
Comment (MikeNachbaur) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m]: NSCoder doesn't return mutable references.  So it's better, in this case, to be explicit and return a mutableCopy of the decoded object.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m [hunk 2]
```diff
@@ -117,22 +122,24 @@ - (void)encodeWithCoder:(NSCoder*)encoder {
     [encoder encodeObject:_communities forKey:kUser_COMMUNITIES];
     [encoder encodeObject:_customData forKey:kUser_CUSTOM_DATA];
     [encoder encodeBool:_guestUser forKey:kUser_IS_GUEST_USER];
+    [encoder encodeInteger:_accessRestrictions forKey:kUser_ACCESS_RESTRICTIONS];
```
Comment (MikeNachbaur) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m]: I noticed that accessRestrictions wasn't being persisted.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m [hunk 3]
```diff
@@ -29,17 +29,18 @@
...
+static NSString * const kUser_EMAIL               = @"email";
+static NSString * const kUser_FULL_NAME           = @"fullName";
+static NSString * const kUser_ORGANIZATION_NAME   = @"organizationName";
+static NSString * const kUser_USER_NAME           = @"userName";
+static NSString * const kUser_COMMUNITY_ID        = @"communityId";
+static NSString * const kUser_COMMUNITIES         = @"communities";
+static NSString * const kUser_ID_DATA             = @"idData";
+static NSString * const kUser_CUSTOM_DATA         = @"customData";
+static NSString * const kUser_IS_GUEST_USER       = @"guestUser";
+static NSString * const kUser_ACCESS_RESTRICTIONS = @"guestUser";
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m]: What's ```ACCESS_RESTRICTIONS```?

Comment (MikeNachbaur) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m]: It's one of the properties on the SFUserAccount.h header.  An `SFUserAccountAccessRestriction` enum

Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m]: The value of this key should be something other than ```guestUser```, I presume?

Comment (MikeNachbaur) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccount.m]: Oh dude, I just noticed I had a ⌘Z fail.  Let me update this.  Thank you for catching it!

**Comment by bhariharan** [comment]
Looks good to me. :+1: 

Are any upgrade steps necessary for existing ```Mobile SDK``` apps?

**Comment by MikeNachbaur** [comment]
There shouldn't be, as long as corrupted or invalid data hasn't been added.  Ideally the `SFOAuthCredentials` encodeWithCoder should use `encodeBool:forKey:` for the SFOAuthEncrypted flag, instead of casting a boolean as an NSNumber object value, but I didn't want to introduce any backwards-compatibility problems that would necessitate upgrade problems.

**Comment by bhariharan** [comment]
Thanks! :+1:

**Pull request merged**